src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http
+{
+    public class DateHeaderValueManager : IDisposable
+    {
+        private readonly TimeSpan _timeWithoutRequestsUntilIdle = TimeSpan.FromSeconds(10);
+        private readonly TimeSpan _timerInterval = TimeSpan.FromSeconds(1);
+        private readonly uint _timerTicksWithoutRequestsUntilIdle;
+
+        private volatile string _dateValue;
+        private bool _isDisposed = false;
+        private bool _hadRequestsSinceLastTimerTick = false;
+        private Timer _dateValueTimer;
+        private object _timerLocker = new object();
+        private int _timerTicksSinceLastRequest;
+
+        public DateHeaderValueManager()
+        {
+            _timerTicksWithoutRequestsUntilIdle = (uint)(_timeWithoutRequestsUntilIdle.TotalMilliseconds / _timerInterval.TotalMilliseconds);
+        }
+
+        public string GetDateHeaderValue()
+        {
+            PumpTimer();
+
+            return _dateValue ?? DateTime.UtcNow.ToString("r");
+        }
+        
+        public void Dispose()
+        {
+            lock (_timerLocker)
+            {
+                if (_dateValueTimer != null)
+                {
+                    DisposeTimer();
+                    _isDisposed = true;
+                }
+            }
+        }
+
+        private void PumpTimer()
+        {
+            _hadRequestsSinceLastTimerTick = true;
+
+            if (!_isDisposed && _dateValueTimer == null)
+            {
+                lock (_timerLocker)
+                {
+                    if (!_isDisposed && _dateValueTimer == null)
+                    {
+                        // Immediately assign the date value and start the timer again
+                        _dateValue = DateTime.UtcNow.ToString("r");
+                        _dateValueTimer = new Timer(UpdateDateValue, state: null, dueTime: _timerInterval, period: _timerInterval);
```

Comment:
The libuv timers are implemented in more or less the same way on [linux](https://github.com/libuv/libuv/blob/bb2632b339d86f69f7dd9aa7168e4b46bf4126a5/src/unix/timer.c#L154) and [win](https://github.com/libuv/libuv/blob/bb2632b339d86f69f7dd9aa7168e4b46bf4126a5/src/win/timer.c#L192) as the .net timer; but you'd pay the added overhead of interop'ing the callback and the code for it would be less clear.

As Kestrel uses multiple (maybe?) libuv threads for socket IO then as you say, this would update in different cpu caches. However, as its a write once per second, read many times the processor cache would only have to refresh once per second (or once every 100k? executions) so it wouldn't be that significant; and it would put extra book keeping on the socket IO threads.